### PR TITLE
Fix losetup / mount

### PR DESCRIPTION
**Pull request by claudiozz**
Summary:

    Split losetup in binary/library to manage loop devices
    Fix losetup to use RDWR when possible and use library
    Fix options in mount
    Add loop option in mount

### CODE REVIEW

File: cmds/losetup/losetup.go [hunk 1]
```diff
@@ -15,90 +15,57 @@ package main
...
-	LOOP_CLR_FD       = 0x4C01
-	LOOP_SET_FD       = 0x4C00
-	LO_NAME_SIZE      = 64
-	LO_KEY_SIZE       = 32
-	/* /dev/loop-control interface */
-	LOOP_CTL_ADD      = 0x4C80
-	LOOP_CTL_REMOVE   = 0x4C81
-	LOOP_CTL_GET_FREE = 0x4C82
-	SYS_ioctl         = 16
+	. "github.com/u-root/u-root/pkg/losetup"
```
Comment (hugelgupf) [inline cmds/losetup/losetup.go]: Please don't use . imports -- they are reserved for some special cases. (I've never actually seen them used for a good reason.)

File: pkg/losetup/losetup.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	defer file.Close()
+
+	device, err := os.OpenFile(devicename, mode, 0644)
+	if err != nil {
+		log.Fatal("Error opening device", devicename)
+	}
+	defer device.Close()
+
+	return LoopSetFd(device.Fd(), file.Fd())
+
```
Comment (hugelgupf) [inline pkg/losetup/losetup.go]: nit: remove empty line?

File: pkg/losetup/losetup.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+	if number, err := LoopCtlGetFree(uintptr(cfd)); err != nil {
+		return "", err
+	} else {
+		return fmt.Sprintf("/dev/loop%d", number), nil
+	}
+
+}
+
+func LoopClearFd(fd uintptr) error {
+	if _, _, err := syscall.Syscall(SYS_ioctl, fd, LOOP_CLR_FD, 0); err != 0 {
```
Comment (hugelgupf) [inline pkg/losetup/losetup.go]: s/SYS_ioctl/syscall.SYS_IOCTL

File: pkg/losetup/losetup.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
+package losetup
```
Comment (hugelgupf) [inline pkg/losetup/losetup.go]: Mind renaming the file to losetup_linux.go?

File: pkg/losetup/losetup.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+func LoopSetFdFiles(devicename, filename string) error {
+	mode := os.O_RDWR
+	file, err := os.OpenFile(filename, mode, 0644)
+	if err != nil {
+		mode = os.O_RDONLY
+		file, err = os.OpenFile(filename, mode, 0644)
+		if err != nil {
+			log.Fatal("Error opening file", filename)
```
Comment (hugelgupf) [inline pkg/losetup/losetup.go]: return error

File: pkg/losetup/losetup.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+		}
+	}
+	defer file.Close()
+
+	device, err := os.OpenFile(devicename, mode, 0644)
+	if err != nil {
+		log.Fatal("Error opening device", devicename)
```
Comment (hugelgupf) [inline pkg/losetup/losetup.go]: return error

File: pkg/losetup/losetup.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+	LO_KEY_SIZE       = 32
+	/* /dev/loop-control interface */
+	LOOP_CTL_ADD      = 0x4C80
+	LOOP_CTL_REMOVE   = 0x4C81
+	LOOP_CTL_GET_FREE = 0x4C82
+	SYS_ioctl         = 16
+)
+
+// consider making this a goroutine which pushes the string down the channel.
+func FindLoopDevice() (name string, err error) {
```
Comment (hugelgupf) [inline pkg/losetup/losetup.go]: super go nit: exported functions should have comments that begin with the identifier and are full sentences for godocs.

```go
// FindLoopDevice finds an unused loop device.
//
// TODO: consider...
```

File: cmds/mount/mount.go [hunk 8]
```diff
@@ -14,31 +14,78 @@ package main
...
+func init() {
+	flag.Var(&options, "o", "Comma separated list of mount options")
+}
+
+func loop(filename *string) error {
+	loopDevice, err := FindLoopDevice()
+	if err != nil {
+		return err
+	}
+	if err := LoopSetFdFiles(loopDevice, *filename); err != nil {
```
Comment (hugelgupf) [inline cmds/mount/mount.go]: Why is filename a pointer? Why take a string and return a string?

File: cmds/losetup/losetup.go [hunk 9]
```diff
@@ -15,90 +15,57 @@ package main
...
 	}
-	var file, dev string
-	var err error
+
 	if len(args) == 1 {
-		dev, err = findloop()
+		devicename, err = FindLoopDevice()
 		if err != nil {
-			l.Fatalf("can't find a loop: %v\n", err)
+			log.Fatalf("can't find a loop: %v\n", err)
```
Comment (hugelgupf) [inline cmds/losetup/losetup.go]: new line not necessary in logs

File: pkg/losetup/losetup.go (hunk 7 above)
Comment (insomniacslk) [reply pkg/losetup/losetup.go]: What about adding `gofmt` / `golint` to commit hooks (or CI tests) to avoid having to catch these things manually?

Comment (hugelgupf) [reply pkg/losetup/losetup.go]: Yeah, I had opened an issue about that a while ago, but golint currently finds ~300 issues in our code base. Additionally, golint is advisory - some of its advice I'd ignore in certain situations. (I'm wondering if there is a flag to turn off a type of advice...)

Comment (insomniacslk) [reply pkg/losetup/losetup.go]: Apparently not :( https://github.com/golang/lint/issues/263

File: pkg/losetup/losetup_linux.go [hunk 10]
```diff
@@ -0,0 +1,107 @@
+package losetup
+
+import (
+	"fmt"
+	"os"
+	"syscall"
+)
+
+const (
+	/*
```
Comment (rminnich) [inline pkg/losetup/losetup_linux.go]: is it worth checking to see if these are now defined in some extant go package? They did not used to be but ...

Comment (claudiozz) [reply pkg/losetup/losetup_linux.go]: I could not find them, any pointers on where to look ?

Comment (claudiozz) [reply pkg/losetup/losetup_linux.go]: Actually you are right the SYS_IOCTL cost is already present in syscall I need to change that.

**Comment by claudiozz** [comment]
Since I foobared my fork I'm closing this pull request and opening https://github.com/u-root/u-root/pull/587

**Pull request closed without merging**